Sources/NIOPosix/PooledRecvBufferAllocator.swift

```diff
@@ -0,0 +1,198 @@
+        if newCapacity > self.capacity {
+            self.capacity = newCapacity
+            self.buffers.reserveCapacity(newCapacity)
+        } else if newCapacity < self.capacity {
+            self.capacity = newCapacity
+
+            // Drop buffers if over capacity.
+            while self.buffers.count > self.capacity {
+                self.buffers.removeLast()
+            }
+
+            // Reset the index if it's too high.
+            if self.index >= self.capacity {
+                self.index = self.buffers.startIndex
+            }
+        }
+    }
+
+    /// Record the number of bytes which were read.
+    ///
+    /// Returns whether the next buffer will be larger than the last.
+    mutating func record(actualReadBytes: Int) -> Bool {
+        return self.recvAllocator.record(actualReadBytes: actualReadBytes)
+    }
+
+    /// Provides a buffer with enough writable capacity as determined by the underlying
+    /// receive allocator to the given closure.
+    @inlinable
+    mutating func buffer<Result>(
+        allocator: ByteBufferAllocator,
+        _ body: (inout ByteBuffer) throws -> Result
+    ) rethrows -> (ByteBuffer, Result) {
+        let index: Int
+
+        if self.isFilling, self.index == self.buffers.endIndex {
+            self.buffers.append(self.recvAllocator.buffer(allocator: allocator))
+            index = self.nextIndex()
+        } else {
+            index = self.nextIndex()
+            let nextBufferSize = self.recvAllocator.nextBufferSize()
+            self.buffers[index].clear(minimumCapacity: nextBufferSize)
+        }
+
+        let result = try body(&self.buffers[index])
+        return (self.buffers[index], result)
+    }
+}
+
+internal struct RecvBufferAllocator {
+    @usableFromInline
+    var _backing: _Backing
+
+    init(_ allocator: RecvByteBufferAllocator, capacity: Int) {
+        self._backing = .init(allocator, capacity: capacity)
+    }
+
+    mutating func updateRecvAllocator(_ allocator: RecvByteBufferAllocator, capacity: Int) {
+        self._backing = .init(allocator, capacity: capacity)
+    }
+
+    mutating func updatePoolCapacity(to newCapacity: Int) {
+        self._backing.updatePoolCapacity(to: newCapacity)
+    }
+
+    mutating func record(actualReadBytes: Int) -> Bool {
+        return self._backing.record(actualReadBytes: actualReadBytes)
+    }
+
+    @inlinable
+    mutating func buffer<Result>(allocator: ByteBufferAllocator,
+                                 _ body: (inout ByteBuffer) throws -> Result) rethrows -> (ByteBuffer, Result) {
+        return try self._backing.buffer(allocator: allocator, body)
+    }
+}
+
+extension RecvBufferAllocator {
+    @usableFromInline
+    enum _Backing {
+        case pooled(PooledRecvBufferAllocator)
+        case unpooled(RecvByteBufferAllocator)
+        case modifying
+
+        init(_ allocator: RecvByteBufferAllocator, capacity: Int) {
+            if let poolable = allocator as? NIOPoolableRecvByteBufferAllocator {
```

Comment:
This is pretty painfully expensive. Perhaps we should amend `RecvByteBufferAllocator` to include the new method with a default implementation that returns `nil`, and have a different pooling strategy for the fallback case?